### ISSUE TITLE
"Most recent" branch source for auto versioning

### ISSUE THREAD

**Issue by dcoraboeuf** [nemerosa/ontrack#1169]
Use case:

Two branches (release/1.1 & release/1.2) are available for a project which is dependency of an auto versioned parent project with the following default branch source:

```yaml
branch: 'release\/1\..*'
```

In this scenario, no promotion has been granted yet in release 1.2 of the dependency.

When 1.1 is promoted, Ontrack identifies a branch on the parent project to be a potential candidate for auto versioning.

This branch is configured to accept only the latest `release/1.*` branch, which is - now - the `release/1.2`.

Therefore, a 1.1 promotion is no longer eligible as soon as the 1.2 branch was created (and registered in Ontrack).

What exactly do we want to achieve? In this scenario, we always want the version promoted in 1.1 as long as there is none in 1.2. Let's imagine we promote a 1.1 while 1.2 was already promoted, what then? How do we protect ourselves?

An idea would be to accept a promotion as long as there is no such a promotion in later branches.

* a 1.1 is promoted and there is no such promotion in more recent branches (1.2, etc.) - we accept it
* a 1.1 is promoted and there is already such a promotion in a more recent branch (1.2 for example) - we reject it

This would mean that we'd have to replace the existing simple branch source:

```yaml
branch: 'release\/1\..*'
```

by:

```
branch: '&most-recent:release\/1\..*'
```

This `most-recent` branch source (name to be defined) does not exist yet and would need to be implemented. The "most recent" means here: accept a promotion if and only if there is no build with the same promotion in a more recent version.

**Comment by dcoraboeuf** [comment]
Implementation notes:

* the `BranchSource` interface would need to know which promotion we target (we already have it during the processing)
* the new `BranchSource` for the "most recent" will need to query the project on the builds with promotions, sorted by the "natural branch order" (using the branch release/version). The latest branch is then the one of the latest promoted build.

**Comment by dcoraboeuf** [comment]
Available in 4.7.31
